File: ios/Classes/PdfMerger.swift
import PDFKit
import UIKit

enum PdfMergerError: Error {
    case notFileUrl
    case fileNotFound
    case writeFailed

    case emptyFilesPath
}

class PdfMerger {

    func choosePagesIndexToMerge(
        inputPath: String,
        outputPath: String,
        pagesIndex: [Int],
        onProgress: ((_ current: Int, _ total: Int) -> Void)? = nil
    ) throws -> String {
        let url = URL(fileURLWithPath: inputPath)
        guard url.isFileURL else {
            throw PdfMergerError.notFileUrl
        }
        let pdfDocument = PDFDocument(url: url)
        guard pdfDocument != nil else {
            throw PdfMergerError.fileNotFound
        }

        let pdfDocumentToMerge = PDFDocument.init()
        pagesIndex.reversed().forEach {
            if let page = pdfDocument?.page(at: $0) {
                pdfDocumentToMerge.insert(page, at: 0)
                onProgress?($0 + 1, pagesIndex.count)
            }
        }
        let success = pdfDocumentToMerge.write(toFile: outputPath)
        guard success else {
            throw PdfMergerError.writeFailed
        }
        return outputPath
    }

    func mergePdfFiles(
        filesPath: [String],
        outputPath: String
    ) throws -> String {
        guard !filesPath.isEmpty else {
            throw PdfMergerError.emptyFilesPath
        }
        let mergedPDF = PDFDocument()
        var pageIndex = 0
        for path in filesPath {
            guard let pdfDocument = PDFDocument(url: URL(fileURLWithPath: path)) else { continue }

            for index in 0..<pdfDocument.pageCount {
                guard let page = pdfDocument.page(at: index) else { continue }
                mergedPDF.insert(page, at: pageIndex)
                pageIndex += 1
            }
        }
        guard mergedPDF.write(to: URL(fileURLWithPath: outputPath)) else {
            throw PdfMergerError.writeFailed
        }
        return outputPath
    }

    func imagesToPdf(
        imagesPath: [String],
        outputPath: String,
        config: ImagesToPdfConfig?
    ) throws -> String {
        return try ImagesToPdfHelper.imagesToPdf(imagesPath: imagesPath, outputPath: outputPath, config: config ?? ImagesToPdfConfig.defaultConfig)
    }
}

File: android/src/main/kotlin/com/example/pdf_worker/PdfMerger.kt
package com.example.pdf_worker

import com.tom_roush.pdfbox.pdmodel.PDDocument
import java.io.File
import java.io.FileNotFoundException

class PdfMerger {
    fun choosePagesIndexToMerge(
            inputPath: String,
            outputPath: String,
            pagesIndex: List<Int>,
            onProgress: ((current: Int, total: Int) -> Unit)? = null
    ): String {
        val file = File(inputPath)
        if (!file.exists()) {
            throw FileNotFoundException("File does not exist: $inputPath")
        }
        val sourceDocument = PDDocument.load(file)
        val mergedDocument = PDDocument()

        if (pagesIndex.isEmpty()) {
            throw IllegalArgumentException("Pages index is empty")
        }
        if (pagesIndex.any { it < 0 || it >= sourceDocument.numberOfPages }) {
            throw IllegalArgumentException("Pages index is out of range")
        }
        pagesIndex.forEachIndexed { index, pageNumber ->
            val page = sourceDocument.getPage(pageNumber)
            mergedDocument.addPage(page)
            onProgress?.invoke(index + 1, pagesIndex.size)
        }

        mergedDocument.save(File(outputPath))

        sourceDocument.close()
        mergedDocument.close()
        return outputPath
    }

    fun mergePdfFiles(filesPath: List<String>, outputPath: String): String {
        val files = filesPath.map { File(it) }
        if (files.any { !it.exists() }) {
            throw FileNotFoundException("File does not exist: ${filesPath.joinToString(", ")}")
        }
        val sourceDocuments = files.map { PDDocument.load(it) }
        val mergedDocument = PDDocument()
        sourceDocuments.forEach { doc ->
            val pages = doc.pages // 这是 PDPageTree
            pages.forEach { page -> mergedDocument.addPage(page) }
        }
        mergedDocument.save(File(outputPath))
        sourceDocuments.forEach { it.close() }
        mergedDocument.close()
        return outputPath
    }

    fun mergeImagesToPdf(
        imagesPath: List<String>,
        outputPath: String,
        config: ImagesToPdfConfig? = null,
    ): String {
        return ImagesToPdfHelper.imagesToPdf(imagesPath, outputPath, config)
    }
}
